### PR TITLE
Allow to load marks in threadpool in advance

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to load marks with threadpool in advance. Regulated by setting `load_marks_asynchronously` (default: 0).

### CODE REVIEW

**Comment by tavplubix** [comment]
Btw, we have a bug in marks loading: https://github.com/ClickHouse/ClickHouse/issues/40399

**Review by CurtizJ — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeMarksLoader.cpp [hunk 1]
```diff
@@ -33,12 +43,49 @@ MergeTreeMarksLoader::MergeTreeMarksLoader(
...
 {
+    if (read_settings_.load_marks_asynchronously)
+    {
+        future = loadMarksAsync();
+    }
+}
+
+MergeTreeMarksLoader::~MergeTreeMarksLoader()
+{
+    if (future.valid())
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeMarksLoader.cpp:54]: Interesting to check, at what moment in time, the future becomes invalid. Looks like some race condition is still possible.

File: programs/local/LocalServer.cpp [hunk 2]
```diff
@@ -227,6 +227,10 @@ void LocalServer::cleanup()
             global_context.reset();
         }
 
+        /// thread status should be destructed before shared context because it relies on process list.
+        thread_status.reset();
+        shared_context.reset();
```
Comment (tavplubix) [inline programs/local/LocalServer.cpp:232]: Consider just placing `thread_status` and `shared_context` fields in `ClientBase` in reverse order

Comment (kssenii) [reply programs/local/LocalServer.cpp:232]: I assume they are already in the right order: `shared_context` in order of destruction goes after `thread_status` in ClientBase. Am I wrong?

Comment (tavplubix) [reply programs/local/LocalServer.cpp:232]: Hm, yes, so they will be destructed in the right order without this change (but the comment is still useful, we can move it to `ClientBase.h`)

**Comment by kssenii** [comment]
I merged branch from PR https://github.com/ClickHouse/ClickHouse/pull/41011 into this branch because they have common code (in `threadPoolcCallbackRunner.cpp` and `LocalServer.cpp`), so easier to merge them now, than later.

So need to wait before https://github.com/ClickHouse/ClickHouse/pull/41011 is merged before merging this PR.

**Comment by kssenii** [comment]
Results look good for s3 and not good for local disk :/
So I'll disable this setting it by default.

- results after running benchmark over `hits_100m_obfuscated` with disk `s3` and disabled cache:

1. `load_marks_asynchronously=0`
``` sql
ip-172-31-4-213.eu-west-1.compute.internal :) select event, value / 1000 / 1000  from system.events where event = 'WaitMarksLoadMicroseconds'

SELECT
    event,
    (value / 1000) / 1000
FROM system.events
WHERE event = 'WaitMarksLoadMicroseconds'

Query id: dae62540-d6e0-48c0-b305-afb68489583b

┌─event─────────────────────┬─divide(divide(value, 1000), 1000)─┐
│ WaitMarksLoadMicroseconds │                         92.359618 │
└───────────────────────────┴───────────────────────────────────┘

1 row in set. Elapsed: 0.005 sec.
```
2. `load_marks_asynchronously=1`
```sql
ip-172-31-4-213.eu-west-1.compute.internal :) select event, value / 1000 / 1000  from system.events where event = 'WaitMarksLoadMicroseconds'

SELECT
    event,
    (value / 1000) / 1000
FROM system.events
WHERE event = 'WaitMarksLoadMicroseconds'

Query id: 1027f644-5563-4d6f-9ffb-f8dba17f08b9

┌─event─────────────────────┬─divide(divide(value, 1000), 1000)─┐
│ WaitMarksLoadMicroseconds │                         17.028783 │
└───────────────────────────┴───────────────────────────────────┘

1 row in set. Elapsed: 0.004 sec.
```
- results after running benchmark over `hits` with `local` disk:
1. `load_marks_asynchronously=0`
```sql
ip-172-31-4-213.eu-west-1.compute.internal :) select event, value / 1000 / 1000  from system.events where event = 'WaitMarksLoadMicroseconds'

SELECT
    event,
    (value / 1000) / 1000
FROM system.events
WHERE event = 'WaitMarksLoadMicroseconds'

Query id: 4146b46b-00bf-4728-9a91-8ed01b23be9d

┌─event─────────────────────┬─divide(divide(value, 1000), 1000)─┐
│ WaitMarksLoadMicroseconds │                          1.628356 │
└───────────────────────────┴───────────────────────────────────┘

1 row in set. Elapsed: 0.004 sec.
```
2. `load_marks_asynchronously=1`
```sql
ip-172-31-4-213.eu-west-1.compute.internal :) select event, value / 1000 / 1000  from system.events where event = 'WaitMarksLoadMicroseconds'

SELECT
    event,
    (value / 1000) / 1000
FROM system.events
WHERE event = 'WaitMarksLoadMicroseconds'

Query id: e59ebb10-f003-4c99-ac3d-3483340b03a7

┌─event─────────────────────┬─divide(divide(value, 1000), 1000)─┐
│ WaitMarksLoadMicroseconds │                4.8989080000000005 │
└───────────────────────────┴───────────────────────────────────┘

1 row in set. Elapsed: 0.130 sec.
```

**Comment by kssenii** [comment]
> Integration tests (tsan) [3/4] — fail: 1

test_dns_cache/test.py::test_host_is_drop_from_cache_after_consecutive_failures - flaky in master.

> Stateless tests (debug, s3 storage) [1/3] — fail: 1

00632_get_sample_block_cache - flaky in master

> Stateless tests (release, s3 storage) — fail: 1

01825_type_json_ghdata_insert_select - DB::Exception: Message: The specified key does not exist. - flaky in master.

**Comment by tavplubix** [comment]
New test 02417_load_marks_async is flaky:
https://s3.amazonaws.com/clickhouse-test-reports/41263/f2acb53887807ce0d1b681a455b0be77589500d1/stateless_tests__aarch64_.html

**Pull request merged**